### PR TITLE
`solveset_complex(sinh(x), x)` returns `ImageSet(Lambda(n, n*I*pi), Integers)`

### PR DESCRIPTION
**Pull request by gxyd**
Fixes issue #9606 
@hargup @aktech

### CODE REVIEW

**Comment by hargup** [comment]
This approach won't work if the given equation has more than one hyperbolic monomials, say `sinh(x) + cosh(x)`. A better approch would be to do something similar to what `solveset_real` does for trignometric equations.

**Comment by gxyd** [comment]
On this [line](https://github.com/sympy/sympy/blob/master/sympy/solvers/solveset.py#L493), why do we use `.subs(exp(I*symbol), y)` for both the `trigonometry` and `hyperbolic`. Should not there be separate for the _solve_real_trig and _solve_real_hyper ? Since currently even the basic real-valued equations like
```
>>> x = Symbol('x', real=True)
>>> solveset(sinh(x), x)
NotImplementedError     # raised error
```
But if we have a separate for both trig and hyperb, would not that be better ?
Since in hyper we should use .subs(exp(symbol), y) not .subs(exp(I*symbol), y).

File: sympy/solvers/solveset.py [hunk 1]
```diff
@@ -776,6 +778,28 @@ def solveset_complex(f, symbol):
         return result
 
 
+def _solve_complex_hyper(f, symbol):
```
Comment (hargup) [inline sympy/solvers/solveset.py]: ` _solve_complex_hyper` -> `_solve_complex_hyperbolic`

Comment (hargup) [inline sympy/solvers/solveset.py]: You don't need to do all these simplifications, breaking up things into fractions and all, simply rewriting the equation into `exp` should be enough.

**Comment by hargup** [comment]
We are using the same function because the basic idea to solve both the types of equations is same "rewrite in exp" and the `_solve_real_trig` function should simply be:
```python
Okay I got it, you wrote this looking at the `_solve_real_trig` function and in theory it should be just
```python
def _solve_real_trig(f, symbol):
    """ Helper to solve trigonometric equations """
    g = f.rewrite(exp)
    return solveset_complex(g, symbol)
```
unfortunately it is not working out of the box and `solveset_complex` is failing on equations like `-I*(exp(I*x) - exp(-I*x))/2 - 1` basically because the rational function is not able to see the expression as a rational function in `exp(I*x)`. Some tweeking in `_solve_as_poly` should fix it.

**Comment by gxyd** [comment]
One of the commits fixes the issue #9759 . I think the problem was not in the `not being able to recognise as polynomial of exp(I*x)` but in that, if `two roots` were same of any `equation`, then it `raised NotImplementedError`.

File: sympy/solvers/tests/test_solveset.py [hunk 2]
```diff
@@ -676,15 +686,13 @@ def test_solveset_complex_tan():
...
     assert solveset_real(sin(x) - 1, x) == \
         imageset(Lambda(n, 2*pi*n + pi/2), S.Integers)
 
     assert solveset_real(cos(x), x) == \
-        Union(imageset(Lambda(n, 2*pi*n - pi/2), S.Integers),
-              imageset(Lambda(n, 2*pi*n + pi/2), S.Integers))
+        imageset(Lambda(n, pi*n + pi/2), S.Integers)
 
     assert solveset_real(sin(x) + cos(x), x) == \
         Union(imageset(Lambda(n, 2*n*pi - pi/4), S.Integers),
```
Comment (gxyd) [inline sympy/solvers/tests/test_solveset.py]: Though this tests fails. On my commits
```
>>> solveset_real(sin(x) + cos(x), x)
ImageSet(Lambda(_n, 2*_n*pi + atan(im(sqrt(-1/(-1 + I))*sqrt(-1 - I))/re(sqrt(-1/(-1 + I))*sqrt(-1 - I))) + atan(im(sqrt(-sqrt(-1 - I))*sqrt(-sqrt(1 + I))*sqrt(-sqrt(-sqrt(2) + 2) - I*sqrt(sqrt(2) + 2))*sqrt(-sqrt(sqrt(2) + 2) + I*sqrt(-sqrt(2) + 2)))/re(sqrt(-sqrt(-1 - I))*sqrt(-sqrt(1 + I))*sqrt(-sqrt(-sqrt(2) + 2) - I*sqrt(sqrt(2) + 2))*sqrt(-sqrt(sqrt(2) + 2) + I*sqrt(-sqrt(2) + 2))))), Integers()) U ImageSet(Lambda(_n, 2*_n*pi + atan(im(sqrt(-1/(-1 + I))*sqrt(-1 - I))/re(sqrt(-1/(-1 + I))*sqrt(-1 - I))) + atan(im(sqrt(-sqrt(-1 - I))*sqrt(-sqrt(1 + I))*sqrt(-sqrt(-sqrt(2) + 2) - I*sqrt(sqrt(2) + 2))*sqrt(-sqrt(sqrt(2) + 2) + I*sqrt(-sqrt(2) + 2)))/re(sqrt(-sqrt(-1 - I))*sqrt(-sqrt(1 + I))*sqrt(-sqrt(-sqrt(2) + 2) - I*sqrt(sqrt(2) + 2))*sqrt(-sqrt(sqrt(2) + 2) + I*sqrt(-sqrt(2) + 2)))) + pi), Integers())
```
which is a nasty big expression.

File: sympy/solvers/solveset.py [hunk 3]
```diff
@@ -776,6 +762,12 @@ def solveset_complex(f, symbol):
         return result
 
 
+def _solve_complex_hyperbolic(f, symbol):
```
Comment (hargup) [inline sympy/solvers/solveset.py]: Here `_solve_complex_hyperbolic` is exactly same as `_solve_real_trig`, I suggest removing `_solve_complex_hyperbolic` and renaming `_solve_real_trig` as `rewrite_to_exp` or maybe `solve_trig_hyperbolic`.

**Comment by hargup** [comment]
@gxyd you should skip fixing https://github.com/sympy/sympy/issues/9759 in this PR, your comment suggest this might need more structral changes in solveset, this is something I've been thinking about and I'll make a post on the mailling list soon. If there are any failing tests, `@xfail` them and add a reference to #9759 or this PR.

**Comment by gxyd** [comment]
@hargup  now i have marked some of the failing one's as `@XFAIL`.
> @gxyd you should skip fixing #9759 in this PR,

Ok.

**Comment by hargup** [comment]
Looks good, +1 when tests pass.

**Comment by gxyd** [comment]
@hargup What to do with the failing `doctest` ? Modify it to `solveset_real(sin(x), x)` ? Instead of `solveset_real(sin(x) - 1, x)`

**Comment by hargup** [comment]
> @hargup What to do with the failing `doctest` ? Modify it to `solveset_real(sin(x), x)` ? Instead of `solveset_real(sin(x) - 1, x)`

Yes that would be good.

File: sympy/solvers/tests/test_solveset.py [hunk 4]
```diff
@@ -699,17 +701,18 @@ def test_solve_invalid_sol():
...
 
 @XFAIL
 def test_solve_trig_simplified():
     from sympy.abc import n
-    assert solveset_real(sin(x), x) == \
-        imageset(Lambda(n, n*pi), S.Integers)
-
-    assert solveset_real(cos(x), x) == \
-        imageset(Lambda(n, n*pi + pi/2), S.Integers)
+
```
Comment (aktech) [inline sympy/solvers/tests/test_solveset.py]: You have some trailing whitespace here.

Comment (gxyd) [inline sympy/solvers/tests/test_solveset.py]: yes thanks.

**Comment by gxyd** [comment]
@hargup i had added the `4th commit` later on. Please have a look at that. It passes the previously `non-passing` doctest. i.e `solveset_real(sin(x) - 1, x)`

**Comment by gxyd** [comment]
@hargup

**Comment by gxyd** [comment]
@hargup I am having trouble to merge this to the master. It contains a lot of consecutive changes to the almost same code. Should i `squash commits` and `push`?

**Comment by aktech** [comment]
@gxyd 

Pull request has been sent to you [#1](https://github.com/gxyd/sympy/pull/1)

**Comment by gxyd** [comment]
Thanks @aktech

**Comment by gxyd** [comment]
@aktech Can you please again resolve the merge conflicts.(its been happening to this PR)
ping @hargup @aktech  Is anything to be done here?

**Comment by aktech** [comment]
@gxyd See [#2](https://github.com/gxyd/sympy/pull/2)

**Comment by aktech** [comment]
BTW, this PR make solving of some simple expressions complicated, compared to current output.

```python
In [4]: solveset(sin(x) + cos(x), x, S.Reals)
Out[4]: ImageSet(Lambda(_n, 2*_n*pi + atan(im(sqrt(-1/(-1 + I))*sqrt(-1 - I))/re(sqrt(-1/(-1 + I))*sqrt(-1 - I))) + atan(im(sqrt(-sqrt(-1 - I))*sqrt(-sqrt(1 + I))*sqrt(-sqrt(-sqrt(2) + 2) - I*sqrt(sqrt(2) + 2))*sqrt(-sqrt(sqrt(2) + 2) + I*sqrt(-sqrt(2) + 2)))/re(sqrt(-sqrt(-1 - I))*sqrt(-sqrt(1 + I))*sqrt(-sqrt(-sqrt(2) + 2) - I*sqrt(sqrt(2) + 2))*sqrt(-sqrt(sqrt(2) + 2) + I*sqrt(-sqrt(2) + 2))))), Integers()) U ImageSet(Lambda(_n, 2*_n*pi + atan(im(sqrt(-1/(-1 + I))*sqrt(-1 - I))/re(sqrt(-1/(-1 + I))*sqrt(-1 - I))) + atan(im(sqrt(-sqrt(-1 - I))*sqrt(-sqrt(1 + I))*sqrt(-sqrt(-sqrt(2) + 2) - I*sqrt(sqrt(2) + 2))*sqrt(-sqrt(sqrt(2) + 2) + I*sqrt(-sqrt(2) + 2)))/re(sqrt(-sqrt(-1 - I))*sqrt(-sqrt(1 + I))*sqrt(-sqrt(-sqrt(2) + 2) - I*sqrt(sqrt(2) + 2))*sqrt(-sqrt(sqrt(2) + 2) + I*sqrt(-sqrt(2) + 2)))) + pi), Integers())
```

Currently:

```
In [5]: solveset(sin(x) + cos(x), x, S.Reals)
Out[5]: ImageSet(Lambda(_n, 2*_n*pi + 3*pi/4), Integers()) U ImageSet(Lambda(_n, 2*_n*pi - pi/4), Integers())
```

Though, I have not checked the correctness.

**Comment by asmeurer** [comment]
Wouldn't hyperbolic solving be simplest via `sinh(x) = -I*sin(I*x)`?

**Pull request closed without merging**